### PR TITLE
Giphy search

### PR DESCRIPTION
**Pull request by mmaxim**
Giphy search for chat.

**Go**
1. Add a new package `giphy` that implements a variety of operations against giphy.com, such as `Search`, `Asset`, and `ProxyURL`.
2. Add a new proxy policy for interacting with Giphy in any way. We run a TCP proxy on the server, and route all traffic to Giphy through it. This allows us to keep our IP secret from Giphy, and to keep the content secret from our server. It mimics the protocol described here: https://signal.org/blog/giphy-experiment/.
3. Add ability for `AttachmentURLSrv` to serve Giphy links.
4. Add a new feature to the `commands` package, allowing for commands to "preview" themselves if they are currently in the input bar. The JS will let the service know the contents of the input bar, and the service can preview the relevant command using the chat UI. In this case, we we use the new Giphy search results UI function to communicate to the frontend if Giphy previews are available.
5. Add the Giphy command which posts a random GIF on send (the only thing that happens for mobile), and also allow the user to select a GIF from a preview set.
6. Gate to just admins for now until we get a real API key.

**JS**
@keybase/react-hackers 
1. Add new Giphy preview component for desktop. It shows a grid of previews the user can click on to send the GIF. 
2. Add a new debounced call into the service to update the contents of the input bar, and make it separate from the typing notifications RPC. The service uses the contents of the input bar to generate the previews, and it should really be debounced to work properly. I'd also like to use this for more stuff, like saving message drafts and prefetching unfurl URLs. The current typing notifications RPC is throttled and is not a good fit for this purpose.
3. Hook all the chat UI stuff up.

Here is a preview of what it looks like on desktop:

image

cc @malgorithms @adamjspooner @cecileboucheron

### CODE REVIEW

File: go/chat/commands/giphy.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+func (s *Giphy) Execute(ctx context.Context, uid gregor1.UID, convID chat1.ConversationID,
+	tlfName, text string) (err error) {
+	if !s.Match(ctx, text) {
+		return ErrInvalidCommand
+	}
+	results, err := s.searcher.Search(libkb.NewMetaContext(ctx, s.G().ExternalG()), s.getQuery(text),
+		s.G().AttachmentURLSrv)
+	if err != nil {
+		s.Debug(ctx, "Execute: failed to get Giphy results: %s", err)
+		return
```
[the remark this answers is not in the document]
Comment (mmaxim) [reply go/chat/commands/giphy.go]: Think its just a merge error, I never do this on purpose

File: go/chat/commands/giphy.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	if s.G().GetAppType() == libkb.MobileAppType {
+		// no preview on mobile yet
+		return
+	}
+	s.Lock()
+	if s.currentOpCancelFn != nil {
+		s.currentOpCancelFn()
+	}
+	select {
+	case <-s.currentOpDoneCb:
```
[the remark this answers is not in the document]
Comment (mmaxim) [reply go/chat/commands/giphy.go]: If it is `nil` we just trigger the `default` case and charge through, which is what we want.

File: shared/chat/conversation/giphy/index.desktop.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                    url={p.previewUrl}
+                    width={scaledWidth(p.previewWidth)}
+                  />
+                </Kb.Box>
+              </Kb.Box2>
+            )
+          })}
+        </Kb.Box2>
+      </Kb.ScrollView>
+      <Kb.Image src={resolveImageAsURL('', poweredByImg)} style={styles.poweredBy} />
```
Comment (buoyad) [inline shared/chat/conversation/giphy/index.desktop.js]: Set `poweredByImg` to this instead

File: shared/chat/conversation/giphy/container.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import {namedConnect} from '../../../util/container'
+import HiddenString from '../../../util/hidden-string'
+import GiphySearch from './index'
+
+type OwnProps = {|conversationIDKey: Types.ConversationIDKey|}
+
+const mapStateToProps = (state, ownProps: OwnProps) => {
+  const {conversationIDKey} = ownProps
+  return {
+    previews: state.chat2.giphyResultMap.getIn([conversationIDKey]) || [],
```
Comment (buoyad) [inline shared/chat/conversation/giphy/container.js]: `getIn([...]) || []` -> `get(..., [])`

File: shared/chat/conversation/messages/wrapper/unfurl/image/video.desktop.js [hunk 5]
```diff
@@ -8,20 +8,20 @@ type State = {
...
-    super(props)
-    this.videoRef = React.createRef()
-  }
   _onClick = () => {
-    if (!(this.videoRef && this.videoRef.current)) {
+    if (this.props.onClick) {
+      this.props.onClick()
+      return
+    }
+    if (!(this._videoRef && this._videoRef.current)) {
```
Comment (buoyad) [inline shared/chat/conversation/messages/wrapper/unfurl/image/video.desktop.js]: no need for `this._videoRef && `

File: shared/chat/conversation/normal/container.js [hunk 6]
```diff
@@ -19,11 +19,14 @@ const mapStateToProps = (state, {conversationIDKey, isPending}) => {
   const meta = Constants.getMeta(state, conversationIDKey)
   const infoPanelOpen = Constants.isInfoPanelOpen(state)
   const isSearching = state.chat2.pendingMode === 'searchingForUsers' && isPending
+  const giphyResults = state.chat2.giphyResultMap.getIn([conversationIDKey])
```
Comment (buoyad) [inline shared/chat/conversation/normal/container.js]: same `getIn` -> `get` change here, can simplify the expression for `showGiphySearch` too

**Pull request merged**